### ISSUE TITLE
Feature - schedule snapshot task adding hours range (anacrontab like)

### ISSUE THREAD

**Issue by MFlyer** [rockstor/rockstor-core#1233]
Thinking about Rockstor running in an office (9to5 or 9to6 like mine, monday to friday/saturday), I thought it could be useful to add new options to limit / better define snapshots task execution:

- Add hours range selection
- Add days range selection

So, when my office it's closed and nobody is working, rockstor doesn't create unuseful snapshots

TO-DO LIST
- [x] js lib + integration (render values)
- [x] handle form on add/modify
- [x] db migration
- [x] add sched task write to rockstortab with new arg crontabwindow
- [x] modify scripts/scheduled_tasks/snapshot & pool scrub to check for crontab window
- [x] add crontab window col to scheduled tasks + beautify values

**Comment by MFlyer** [comment]
Looking for suggestions @schakrava @phillxnet 

**First "crash"**
except for js integration (found useful to create a new plugin starting from cron plugin), got a practical problem: 
the new "task window for execution" param, that is going to be the second arg for snapshots schedule, accept start time (hh:mm) - end time (hh:mm) - start and end date (dayofweek)
Avail options i thought to add

- "Always" (no time / days limits)
- "Time window" (no day limits / time limits to from)
- "Days window" (no time limits / day limits to from)
- "Custom window" (time & day limits, user select for ex. from 09.00 to 17.00 on Mon-Fri)

It's possible to add alternatives, but thinking about it was causing an headache xD (every day from a particular hour without end, every day till an hour, every hour starting from a day or stopping to another one etc etc - maybe i add these 4, i like them! :) but no more )

Time window will apply only for tasks with "Every minutes/hour" cos' it make no sense to limit a task executing every day at 4:53 pm or similar.
Do you agree?

**Comment by schakrava** [comment]
Those 4 sound plenty. Your original user story is to be able to exclude scheduled tasks from running during offhours or weekends. So the UI can provide input for exclusions or inclusions. I can't imagine a good UI design atm, but my first take is that may be you can come up with html design just for 4. If you can come up with a simple design for that, then you can achieve 1-4 with one implementation. Easier said than done, good luck :)

**Comment by MFlyer** [comment]
First running version (UI side) - some samples attached
UI side just missing check for hour/min start < hour/min stop day-start < day-stop

![default](https://cloud.githubusercontent.com/assets/17988902/14144451/0b710d26-f690-11e5-954b-4800e8f9d437.jpg)
![example1](https://cloud.githubusercontent.com/assets/17988902/14144458/15fb7d9e-f690-11e5-8185-dc716f0add91.jpg)
![example2](https://cloud.githubusercontent.com/assets/17988902/14144468/21b6a956-f690-11e5-860e-6208b87e56ae.jpg)

**Comment by phillxnet** [comment]
@MFlyer This is looking great. What do you think about adding the word "inclusive" at the end? E.g.
Week days filter - from - *monday* - to - * friday* - **inclusive**
That way it's obvious that the end time is not an up until but an inclusive.

**Comment by MFlyer** [comment]
@schakrava @phillxnet pls add all your suggestions - in particular for texts, i was arguing with myself for that xD

**Comment by MFlyer** [comment]
![infos](https://cloud.githubusercontent.com/assets/17988902/14180720/9afe5172-f764-11e5-9d59-2c7c633b9adf.jpg)

@phillxnet what do you think about this? :) Adding other text after inputs was making them go on next line

**Comment by phillxnet** [comment]
@MFlyer OK yes if it's messing with the formatting then putting in the mouse over is dandy, probably a better idea anyway to keep the clutter down. All I would suggest is that the tool tip wording be a little less wordy but great. It's nice to see this coming along.

As a general point is it OK to declare jquery-cron-window.js as GPLv2 when it was originally MIT? Sorry not up on re-licence issues, presumably you are re-licencing to cover your modifications there in.
See their: LICENSE file.
Also, and I think we are going to need @schakrava 's input on this one, maybe this library should go in the rockstor-jslibs repo, sorry but haven't yet included another js library to know for sure:
There is a section in the docs "Adding third party Javascript libraries" over at http://rockstor.com/docs/contribute.html#adding-third-party-javascript-libraries
but then it looks like you are making fairly heavy changes anyway. Just a tiny and probably wrong point anyway, I'll leave you be.

**Comment by MFlyer** [comment]
Hi @phillxnet had the same doubts about licenses, but reading original jquery-cron.js there's a double licence (MIT + GPLv2), so... @schakrava what to do???

Third party javascript libraries: forked repo and added file there - commits are from a rockstor js library fork branch ;) - although that was not easy (after a build static no library detected so i had to manually add <script scr etc etc> to add_task.jst for testing

Anyway, tomorrow going on with testing & control functions ( start < stop and possibly auto reduce options. Ex. start time 10:20 stop top selects will be only 10-23:21-59 )

**Comment by MFlyer** [comment]
Some news to @schakrava & @phillxnet 

**Days**: stop day always get auto equal to start day, then user select his day
![dayinit](https://cloud.githubusercontent.com/assets/17988902/14226233/40c44be8-f8dc-11e5-850b-7052c8b8ce95.jpg) ![daytoday](https://cloud.githubusercontent.com/assets/17988902/14226234/40cb2684-f8dc-11e5-9ff5-6be56bd1f341.jpg)

**Hours & Mins**: 
- on start hour change auto stop hour = start hour (if star min already == 59 auto correct stop hour & mins to go next hour min 0)
- on start minute change auto stop minute = start min+1 (if start min == 59 stop hour+1 & stop min 0)

![timeinit](https://cloud.githubusercontent.com/assets/17988902/14226256/56385040-f8dd-11e5-8c38-a2a5216ab30d.jpg)
![timehourchange](https://cloud.githubusercontent.com/assets/17988902/14226264/7d156842-f8dd-11e5-8b7b-246b0c72e27e.jpg)
![timehourchanged](https://cloud.githubusercontent.com/assets/17988902/14226266/8573d410-f8dd-11e5-929c-728db0bd6f27.jpg) _**Stop hour changing on start hour changed**_

![timeminutechanged](https://cloud.githubusercontent.com/assets/17988902/14226272/c0d2f144-f8dd-11e5-9e6b-0b8301c8f078.jpg)_**Stop mins changing to +1 on start mins changed**_

![nexthour](https://cloud.githubusercontent.com/assets/17988902/14226275/e1730b3c-f8dd-11e5-826e-7dee207cc718.jpg)_**Start mins == 59, move 1 min ahead that means next hour min 0**_

Stop hour, mins & days: decided not to add a control to force stops > stars, but just a warning for the users.
So doing users have full possibilites: for example a task from 09.00 to 10.30 on Mon-Fri will be ok without warning and a task from 19.30 to 08.30 on Mon-Fri or Fri-Tue will be always ok but with a warning (something like "Hey guy, you're not using the usual time sequence 00.00 to 23.59 Mon Fry, it will be ok, are you sure what your doing?")

Last check to add: hide task execution windows for task != to minutes or hours (it makes no sense to add it on a task running every 3rd of april at 3.54 or similar)

Flyer/Mirko

**Comment by MFlyer** [comment]
Check this :) [Latest tests & mods video](https://youtu.be/Fhofg-fn_Ik)

**Comment by schakrava** [comment]
Looks very impressive!!!!

Is there way to go one step beyond warning on wrong input? the warning message is great but what happens if they ignore and submit anyway?

Feel free to open the pr soon as you think its ready to be tested to some degree.

regarding js libraries, i'll take care of it during merge time.

**Comment by MFlyer** [comment]
Hi @schakrava, let's see prev video message:

> Stop hour, mins & days: decided not to add a control to force stops > stars, but just a warning for the users.
So doing users have full possibilites: for example a task from 09.00 to 10.30 on Mon-Fri will be ok without warning and a task from 19.30 to 08.30 on Mon-Fri or Fri-Tue will be always ok but with a warning (something like "Hey guy, you're not using the usual time sequence 00.00 to 23.59 Mon Fry, it will be ok, are you sure what your doing?")

We accept every data, you'll be able to add limits for example from 9 to 17 Mon Fri (and we call this "conventional") or from 17 to 6 Sun Tue ("unconventional" time format, because hours mins days aren't clockwise).
First case snapshots get created during 9-17 and every day from mon to fri, second case snapshots get created only if it's >17 and <6am only from sunday to tuesday.
That one showing it's just a warning that sounds like "Hey, look what are you doing, if that is your desired timinig ok, else change it, but we accept it in this format too"

Info needed to go on: i'm going to pr js library today (think so, just some checks), than i'll work on rockstore/core to manage data post, snapshots scripts control etc etc.

What about models??? Finally I've decided to add a field to rockstor db and not to use the same field of crontab (2 different fields = clear schema) and have to perform a migration (VM still on 12-05). What's is better? VM Update to latest version and then coding or just don't care about new version and perform db migration?

Flyer

**Comment by MFlyer** [comment]
to @schakrava - Coding stop on migration matter :confused: 

See latest commit for task_def mods (migration task doesn't raise any error) https://github.com/MFlyer/rockstor-core/commit/40e86875c37e5bda6561569c9567fb65ab069ac9

On migration i get this following response and I'm not sure if safe (looked at task model and actually task.start is togheter an index key field and a nullable field - or at least a field that on empty text uses null instead of empty - ?!?, so the auto correction proposed makes sense)

Waiting for you to procede with migration and finishing code :blush: 
```
[root@rockstone models]# /opt/rockstor/bin/django schemamigration smart_manager --auto
 + Added field crontabwindow on smart_manager.TaskDefinition
 ? The field 'Task.start' does not have a default specified, yet is NOT NULL.
 ? Since you are making this field nullable, you MUST specify a default
 ? value to use for existing rows. Would you like to:
 ?  1. Quit now.
 ?  2. Specify a one-off value to use for existing columns now
 ?  3. Disable the backwards migration by raising an exception; you can edit the migration to fix it later
 ? Please select a choice: 3
 ~ Changed field start on smart_manager.Task
Created 0008_auto__add_field_taskdefinition_crontabwindow__chg_field_task_start.py. You can now apply this migration with: ./manage.py migrate smart_manager
```

**Comment by MFlyer** [comment]
Found my way without --auto and this + migrate next
`/opt/rockstor/bin/django schemamigration smart_manager --add-field TaskDefinition.crontabwindow`

![newdb](https://cloud.githubusercontent.com/assets/17988902/14245008/af576d46-fa5e-11e5-81d5-29a8e524d90b.jpg)

@schakrava Now i just need to know if Rockstor still need migration file (to add on git) or not

Going on with code, hope to close it before friday!

**Comment by MFlyer** [comment]
Update on 6 Apr
To do list:
- [x] js lib + integration (render values)
- [x] handle form on add/modify
- [x] db migration
- [x] add sched task write to rockstortab with new arg crontabwindow
- [ ] modify scripts/scheduled_tasks/snapshot & pool scrub to check for crontab window
- [ ] add crontab window col to scheduled tasks + beautify values

**Comment by MFlyer** [comment]
Hi guys @schakrava @phillxnet , got good news: added all we need to limits tasks execution and just missing scheduled tasks page "beautify" to show crontab window execution column

Performed tests with "conventional" time format (ex.: task from 9 to 17 mon - fri), with "unconventional" too (ex. task from 23 to 6 mon - fri or 9 to 15 fri - tue, etc etc) and it seems all feeling good :smiley: 

Note for @schakrava : choosed to add crontabwindow.py as a separate module cos' both snapshot and scrub task share same code (to avoid doubled code!)

Personal consideration: not really sure about utility of crontab window over scrubs (usually you set 1 scrub/day on a specified time), but we've got it too - everyone will be free to decide

Updated first message with task list, hope to close tomorrow, as planned

Flyer/Mirko

P.S.: commits sent testing git new gpg signing option, nice!

**Comment by MFlyer** [comment]
Little preview on tasks list interface ;)

![immagine](https://cloud.githubusercontent.com/assets/17988902/14369631/05aed2f4-fd27-11e5-95c4-639ac49f00f2.jpg)

Finished coding, sent PR

**Comment by MFlyer** [comment]
Branch merged and avail with Rockstor 3.8.13-03 - Issue closed
